clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -849,6 +856,41 @@ class KernelObjVisitor {
   void VisitRecord(const CXXRecordDecl *Owner, ParentTy &Parent,
                    const CXXRecordDecl *Wrapper, Handlers &... handlers);
 
+  // Base case, only calls these when filtered.
```

Comment:
This 'VisitUnion' function is the main trickery.  See this to play around with it: (https://godbolt.org/z/z8b61d) .

There are 3 versions of the function.  This first is the 'base' case.  

The list of 'handlers' not specified via <> syntax (as we do when they aren't specified) are deduced like normal.  This ends up in the normal 'Handlers' type pack, with the first of that pack being picked up by the CurHandler.  

the return type is done with std::enable_if, which switches based on whether we should VisitUnionBody for the 'first' handler (the one in CurHandler).

If we SHOULD visit it, we call VisitUnion again, this time with the CurHandler put into the "FilteredHandlers" pack (which will never be deduced, since it is not the 'last' thing.   
If we SHOULDN'T visit it, we call VIsitUnion again, this time with the CurHandler discarded from the list, and forward on the"FilteredHandlers" (which need to be specified inside <...>, since they cannot be deduced).  

If there are no more items to be put into the cur_handler type (or deduced as a Handlers type), the base is called, which does the work.